### ISSUE TITLE
Cache do not update per second one day

### ISSUE THREAD

**Issue by javalover123** [cache2k/cache2k#195]
Cache do not update per second one day, and there no thread name starts with cache2k-loader-

`
<pre>
Cache<String, LocalDateTime> TIME_CACHE = new Cache2kBuilder<String, LocalDateTime>() {
    }.refreshAheadPolicy(new AutoRefreshPolicy<>())
            .expireAfterWrite(1, TimeUnit.SECONDS)
            .loader((key) -> LocalDateTime.now())
            .build();
</pre>
`

**Comment by cruftex** [comment]
Do you test with the latest SNAPSHOT code? Can you give a complete example of what you are trying with the cache and what you would expect, like a test case? Since you use the code that is currently in progress, it might be unfinished or a bug. Maybe its also a wrong assumption of the functionality.

**Comment by javalover123** [comment]
Yes, use 2.8-SNAPSHOT, we used 4 months, and it's right util this time.

I execute jstack, there saw only 2 cache2k-scheduler threads, then I restart the application and saw another cache2k-loader- thread.

Note: cache expire after write 1 second.
`
TIME_CACHE.get(market)
`

**Comment by cruftex** [comment]
> Yes, use 2.8-SNAPSHOT, we used 4 months, and it's right util this time.

It would be better to use a released version. The SNAPSHOT might have varying quality.

Still, I am not sure what you are doing. Can you give me detailed information and steps how to reproduce your observations?

**Comment by javalover123** [comment]
Below is the simplest test code.

<pre>
import java.time.LocalDateTime;
import java.util.concurrent.TimeUnit;
import org.cache2k.Cache;
import org.cache2k.Cache2kBuilder;
import org.cache2k.expiry.RefreshAheadPolicy;
import org.junit.Test;

/**
 * Cache refresh failed one day. https://github.com/cache2k/cache2k/issues/195
 *
 * @author javalover123
 * @date 2022/10/28
 */
public class Cache2kRefreshTest {

    static Cache<String, LocalDateTime> TIME_CACHE = new Cache2kBuilder<String, LocalDateTime>() {
    }.refreshAheadPolicy(new AutoRefreshPolicy<>())
            .expireAfterWrite(1, TimeUnit.SECONDS)
            .loader((key) -> LocalDateTime.now())
            .build();

    public static class AutoRefreshPolicy<K, V, T> implements RefreshAheadPolicy<K, V, T> {

        @Override
        public long refreshAheadTime(Context<T> ctx) {
            return ctx.getExpiryTime();
        }

        @Override
        public int requiredHits(Context<T> ctx) {
            return 0;
        }

    }

    @Test
    public void refresh() throws InterruptedException {
        LocalDateTime last = null;
        for (int i = 1; i <= 100000; i++) {
            final LocalDateTime now = TIME_CACHE.get("");
            if (now.equals(last)) {
                System.out.println(now + ",refresh error," + last);
            }
            last = now;
            Thread.sleep(1000);
        }
        System.out.println(LocalDateTime.now() + ",refresh done," + last);
    }

}
</pre>

**Comment by cruftex** [comment]
Thanks for the update! How many refresh errors do you get?

**Comment by cruftex** [comment]
I think the cause of your problem is, that by default cache2k is expiring (or refreshing) timer based and execution of timer events will always lag behind. This behavior is saving resources. With typical cache expiry times around 5 minutes, it does not matter if an update happens one second later.

For the purpose of testing you can add `sharpExpiry(true)` to the configuration. Then the test code should pass. See the documentation on `expiryAfterWrite` and `sharpExpiry`.

**Comment by javalover123** [comment]
Not lag minutes, its hours. For example 3pm today execute TIME_CACHE.get("") get yesterday

**Comment by javalover123** [comment]
I execute jstack per 0.5 second, there is a same thread "cache2k-loader-_Cache2kRefreshTest.CLINIT-22-dk-1" #30, and not saw the thread when cache not update.

![image](https://user-images.githubusercontent.com/9191388/198495855-629f154e-09e5-474c-9ae8-2911b3545ef7.png)

**Comment by cruftex** [comment]
What is the git revision of cache2k you are using? What Java runtime version you are using?

**Comment by javalover123** [comment]
git revision is [24 May](https://github.com/cache2k/cache2k/commit/437c58c1ceb0cacaa9d2d282ad702e6aeed3e19d)

Run on windows, jdk is below.
java version "11.0.12" 2021-07-20 LTS
Java(TM) SE Runtime Environment 18.9 (build 11.0.12+8-LTS-237)
Java HotSpot(TM) 64-Bit Server VM 18.9 (build 11.0.12+8-LTS-237, mixed mode)

**Comment by cruftex** [comment]
Ok, I will investigate. 

In the code I am in the middle of some restructuring, however, I was held up by other projects since May. 
The current unreleased code can have issues, since the restructuring is not totally finished yet. I still have around 25 code location where I need to double check things and do corrections.

If this is blocking your project maybe consider using the latest released final version. Before the next release I will do some more changes around the `RefreshAheadPolicy`, which will potentially break your code. Only after the final release I keep the interface stable.

Anyhow, it is great that you test with the latest code base and give feedback. That catches potential problems early on.

**Comment by cruftex** [comment]
> I execute jstack per 0.5 second, there is a same thread "cache2k-loader-_Cache2kRefreshTest.CLINIT-22-dk-1" https://github.com/cache2k/cache2k/issues/30, and not saw the thread when cache not update.

Do I assume correctly that if you start the application its sometimes not working at all? If so, approximately, out of how many starts it does fail?

My strategy is to add a configuration similar to yours to our test and run these tests many times to make sure there is no spurious problem within the initialization procedures.

**Comment by javalover123** [comment]
Run 4 months, just happened once only.

**Comment by ieugen** [comment]
Are you using ECC memory on the servers? 
Random bit flips might also be an issue if not using ECC.

**Comment by javalover123** [comment]
> Are you using ECC memory on the servers? Random bit flips might also be an issue if not using ECC.

We use cloud server, maybe I can't answer these questions.

**Comment by javalover123** [comment]
> Are you using ECC memory on the servers? Random bit flips might also be an issue if not using ECC.

Maybe is non-ECC memory.

![image](https://user-images.githubusercontent.com/9191388/204092416-465782f0-384d-405a-b133-d7c43d92c192.png)

**Comment by ieugen** [comment]
If this is the case and there is no other information we might close the issue.

**Comment by javalover123** [comment]
I execute jstack per 0.5 second, there is a same thread "cache2k-loader-_Cache2kRefreshTest.CLINIT-22-dk-1" #30, and not saw the thread when cache not update.

![image](https://user-images.githubusercontent.com/9191388/198495855-629f154e-09e5-474c-9ae8-2911b3545ef7.png)
